### PR TITLE
Updated AST builder to handle dot notation.

### PR DESCRIPTION
**Pull request by hjafarpour**
### Description 
This PR makes changes in our AST so we can handle dot notation to access nested fields in struct correctly.
Previously we assumed that we have only one dot, stream/table name.column name. How we will be able to have more than one dot in a dereferenced node.
This is based on the previous PR for Query rewrite.
We also check for ambiguity resulting from having columns with the same name as the stream/table name.

### Testing done 
_Describe the testing stategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 1]
```diff
@@ -1164,13 +1195,49 @@ public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
...
+  private boolean isAmbiguousColumnName(
+      final String columnName,
+      final String suffixName,
+      final String sourceName,
+      final String sourceAlias,
+      final Schema sourceSchema
+  ) {
+    boolean isASchemaColumn = sourceSchema.fields().stream()
+        .anyMatch(field -> columnName.equalsIgnoreCase(field.name()));
+    if (suffixName == null) {
```
Comment (rodesai) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: can't we just omit this case? If there is no dereference there is certainly no ambiguity.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 2]
```diff
@@ -1164,13 +1195,49 @@ public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
...
+      return isASchemaColumn
+          && (columnName.equalsIgnoreCase(sourceName) || columnName.equalsIgnoreCase(sourceAlias));
     }
+    return isASchemaColumn
+        && !columnName.equalsIgnoreCase(suffixName)
```
[the remark this answers is not in the document]
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Yes, that's the intended behavior and as you said we should document it an keep it simple. In the above case for instance `S.S` will assume that the first `S` is the stream name and the second `S` is the column name.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 1 above)
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Consider this example:
`CREATE STREAM S (S STRUCT<S INTEGER>) ...;`
`SELECT S FROM S;`
As you see we have ambiguity of the column name is the same as the stream name/alias.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 3]
```diff
@@ -1129,21 +1135,46 @@ public Node visitSubqueryExpression(SqlBaseParser.SubqueryExpressionContext cont
...
+    dotStack.push(fieldName);
+    Expression baseExpression = (Expression) visit(context.base);
+    dotStack.pop();
+    return new DereferenceExpression(getLocation(context), baseExpression, fieldName);
   }
 
   @Override
   public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
     String columnName = getIdentifierText(context.identifier());
+    String suffixFieldName = null;
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: final, (using ? operator).

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 4]
```diff
@@ -1164,13 +1195,49 @@ public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
...
-              QualifiedName.of(dataSourceExtractor.getFromAlias())
-          );
-      return new DereferenceExpression(getLocation(context), baseExpression, columnName);
+      if (isAmbiguousColumnName(
+          columnName,
+          suffixFieldName,
+          dataSourceExtractor.getFromName(),
+          dataSourceExtractor.getFromAlias(),
+          dataSourceExtractor.getFromSchema())) {
+        throw new KsqlException(String.format("Ambiguous column name %s. You have stream/table"
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: yer, just rename it to `throwOnAmbiguousColumnName`

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 5]
```diff
@@ -1129,21 +1135,46 @@ public Node visitSubqueryExpression(SqlBaseParser.SubqueryExpressionContext cont
...
+          dataSourceExtractor.getLeftName(),
+          dataSourceExtractor.getLeftAlias(),
+          dataSourceExtractor.getJoinLeftSchema())
+          || isAmbiguousColumnName(
+          columnName,
+          suffixFieldName,
+          dataSourceExtractor.getRightName(),
+          dataSourceExtractor.getRightAlias(),
+          dataSourceExtractor.getJoinRightSchema())) {
+        throw new KsqlException(String.format("Ambiguous column name %s. You have stream/table"
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Maybe:

> Ambiguous column identifier '%s'. The identifier matches the stream or table identifier. Prefix the identifier with the stream or table identifier to disambiguate.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 1 above)
Comment (big-andy-coates) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: > SELECT S FROM S;

Is there any ambiguity here? Does not the `S` in the select always have to be the struct? It can't be the stream S or the struct's field S...

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Let me just repeat that back so I know I've understood that correctly:

we're saying that if the first part of the column name matches the stream/table name/alias, then we assume that's what it is?

e.g. 
```
CREATE STREAM S (S STRUCT<S INTEGER, C VARCHAR>) ...;
SELECT S.S FROM S;   -- selects whole struct.
SELECT S.C FROM S;   -- fails as there is no first-level column C.
```

I think there is still room for people to grok the first statement wrong. I think most people would expect it to select the integer.

Can we not just fail if the table/stream name/alias matches any first-level column identifier?

File: ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java [hunk 6]
```diff
@@ -1147,7 +1147,7 @@ protected Node visitInsertInto(final InsertInto node, final Object context) {
   protected Node visitDropTable(final DropTable node, final Object context) {
     return node.getLocation()
         .map(location ->
-            new DropTable(node.getLocation().get(),
+            new DropTable(node.getLocation(),
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java]: surely, this style of method can not be simplified to:

```
protected Node visitDropTable(final DropTable node, final Object context) {
    return new DropTable(node.getLocation(),
                         node.getTableName(),
                         node.getIfExists(),
                         node.isDeleteTopic());
  }
```

That was kinda the whole point of making the constructors taking `Optional` public! lol.

File: ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java [hunk 7]
```diff
@@ -717,4 +811,79 @@ public void shouldSetShowDescriptionsForShowQueriesDescriptions() {
     ListQueries listQueries = (ListQueries)statement;
     Assert.assertThat(listQueries.getShowExtended(), is(true));
   }
+
+  @Test
+  public void shouldFailIfStreamColumnNameIsAmbiguous() {
+    final String statementString =
```
Comment (big-andy-coates) [inline ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java]: Why is this ambiguous?

File: ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java [hunk 8]
```diff
@@ -717,4 +811,79 @@ public void shouldSetShowDescriptionsForShowQueriesDescriptions() {
...
+        "CREATE STREAM S AS SELECT address FROM address a;";
+    try {
+      final List<Statement> statements = KSQL_PARSER.buildAst(statementString, metaStore);
+      Assert.fail();
+    } catch (KsqlException e) {
+      assertThat(e.getMessage(), equalTo("Ambiguous column name ADDRESS. You have stream/table with the same name/alias. Use stream/table name or alias as prefix when there is a column with the same name."));
```
Comment (big-andy-coates) [inline ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java]: Use `junit.ExpectedException` - much nicer in your tests.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: made the change.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 1 above)
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: No ambiguity but the idea is to be consistent in the way we treat the cases where we have column name the same as stream name/alias and require user to always provide stream name/alias in referring to such columns.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 9]
```diff
@@ -1129,21 +1135,46 @@ public Node visitSubqueryExpression(SqlBaseParser.SubqueryExpressionContext cont
...
     String fieldName = getIdentifierText(context.identifier());
-    Expression baseExpression;
-    QualifiedName tableName = QualifiedName.of(
-        context.primaryExpression().getText().toUpperCase());
-    baseExpression = new QualifiedNameReference(
-        getLocation(context.primaryExpression()), tableName);
-    DereferenceExpression dereferenceExpression =
-        new DereferenceExpression(getLocation(context), baseExpression, fieldName);
-    return dereferenceExpression;
+    dotStack.push(fieldName);
```
[the remark this answers is not in the document]
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Just did, IntelliJ should make generated local variables and method parameters final from now on, if not I'll blame them ;)

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: @big-andy-coates do you mean fail the ddl statement? Is the reasoning that this is an easier error to deal with then dealing with unexpected successful behavior later on? 

I think if a user wants a stream and one of the columns to have the same name we should let them. If a later column reference has an ambiguity we should have a sane way of resolving it. Detecting if there actually is an ambiguity is a bit complex (especially if you have subscripts). Defaulting to the stream name whenever the first level of the column reference matches the stream name seems ok for now.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java (hunk 6 above)
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java]: Yes, missed this part. Updated it for all the ones that have public constructor now.

File: ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java (hunk 7 above)
Comment (hjafarpour) [reply ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java]: This is based on the above assumption that if you have a column with the same name as stream name/alias you have to always prefix the stream name/alias to use that column.

**Comment by hjafarpour** [comment]
@big-andy-coates actually removed this change and went back to the previous implementation with a minot check which will behave the way you asked. Let me know what you think about it.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 10]
```diff
@@ -1164,12 +1179,21 @@ public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
         throw new InvalidColumnReferenceException("Field " + columnName + " is ambiguous.");
```
Comment (rodesai) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: not exactly related to this patch but the error message here is incorrect. Should be something like ```"Field" + columnName + " is present in neither the left nor right expression"

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 11]
```diff
@@ -1127,23 +1133,32 @@ public Node visitSubqueryExpression(SqlBaseParser.SubqueryExpressionContext cont
...
+    final Expression baseExpression = (Expression) visit(context.base);
+    dotStack.pop();
+    return new DereferenceExpression(getLocation(context), baseExpression, fieldName);
   }
 
   @Override
   public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
-    String columnName = getIdentifierText(context.identifier());
+    final String columnName = getIdentifierText(context.identifier());
+    final String suffixFieldName = dotStack.empty() ? null : dotStack.peek();
```
Comment (rodesai) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: nit: might improve readability later on to just compute a boolean here: ```final boolean hasDereference = !dotStack.empty();```

File: ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java [hunk 12]
```diff
@@ -998,165 +998,82 @@ protected Node visitTableElement(final TableElement node, final Object context)
   }
 
   protected Node visitCreateStream(final CreateStream node, final Object context) {
-    return node.getLocation()
```
Comment (rodesai) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java]: nice!

File: ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java [hunk 13]
```diff
@@ -717,4 +812,77 @@ public void shouldSetShowDescriptionsForShowQueriesDescriptions() {
...
+    assertThat(query.getQueryBody(), instanceOf(QuerySpecification.class));
+    QuerySpecification querySpecification = (QuerySpecification) query.getQueryBody();
+    assertThat(querySpecification.getSelect().getSelectItems().get(0).toString(),
+        equalTo("A.ADDRESS ADDRESS"));
+  }
+
+  @Test
+  public void shouldNotAddPrefixIfStreamNameIsPrefix() {
+    final String statementString =
+        "CREATE STREAM S AS SELECT address.city FROM address a;";
```
Comment (rodesai) [inline ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java]: This test may be less confusing if you use a field thats in the orders schema instead. This is technically not a correct statement. Its juts that the parser isn't responsible for catching what's wrong with it.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 11 above)
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Yes, with the new changes there is no need for `suffixFieldName` anymore.

**Comment by hjafarpour** [comment]
@big-andy-coates this is ready for your review :)

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Hummm... what I'm worried about is allowing syntax that is very easily misinterpreted by the next person that comes along and reads it.  Given:

```
CREATE STREAM S (S STRUCT<S INTEGER, C VARCHAR>) ...;
SELECT S.S FROM S;   -- selects whole struct.
```

Someone coming along with SQL experience, but without in-depth knowledge of the nuances of KSQL, would likely think its selecting the integer field. 

Just documenting "Well, we treat it this way" is all well and good, but it requires users to have read the documentation. And as we all know, not many do!

Hence wondering if just failing on the DLL statement isn't the way to go. With a message like:

> STRUCT column name can not match Stream name or alias. Rename the stream or column. You can use 'CREATE STREAM S_ WITH(KAFKA_TOPIC="S") ...' to keep the underlying topic name the same.

If we later think of a better way to handle this, then we can always allow such statements later. But if we accept them now, its harder to stop later.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Great - but you can go one step further and configure IntelliJ to highlight missing 'final's 🗡 

See: https://confluent.slack.com/archives/C8486MR1C/p1529069775000398

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 14]
```diff
@@ -1127,23 +1133,31 @@ public Node visitSubqueryExpression(SqlBaseParser.SubqueryExpressionContext cont
...
+    return new DereferenceExpression(getLocation(context), baseExpression, fieldName);
   }
 
   @Override
   public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
-    String columnName = getIdentifierText(context.identifier());
+    final String columnName = getIdentifierText(context.identifier());
     // If this is join.
     if (dataSourceExtractor.getJoinLeftSchema() != null) {
+      boolean sameAsLeft = columnName.equalsIgnoreCase(dataSourceExtractor.getLeftAlias())
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: nit: final.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 15]
```diff
@@ -1164,12 +1178,21 @@ public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
...
-      return new DereferenceExpression(getLocation(context), baseExpression, columnName);
+      if (!dotStack.empty()
+          && (columnName.equalsIgnoreCase(dataSourceExtractor.getFromAlias())
+          || columnName.equalsIgnoreCase(dataSourceExtractor.getFromName()))) {
+        return new QualifiedNameReference(
+            getLocation(context),
+            QualifiedName.of(columnName)
+        );
+      } else {
+        Expression baseExpression =
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: nit: final.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 16]
```diff
@@ -1164,12 +1178,21 @@ public Node visitColumnReference(SqlBaseParser.ColumnReferenceContext context) {
...
-          );
-      return new DereferenceExpression(getLocation(context), baseExpression, columnName);
+      if (!dotStack.empty()
+          && (columnName.equalsIgnoreCase(dataSourceExtractor.getFromAlias())
+          || columnName.equalsIgnoreCase(dataSourceExtractor.getFromName()))) {
+        return new QualifiedNameReference(
+            getLocation(context),
+            QualifiedName.of(columnName)
+        );
+      } else {
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: nit: no need for 'else' as you've returned from within the 'if'

File: ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java [hunk 17]
```diff
@@ -998,165 +998,82 @@ protected Node visitTableElement(final TableElement node, final Object context)
...
-                        e -> (Expression) process(e.getValue(), context)
-                    )),
-                node.getPartitionByColumn().isPresent()
-                    ? Optional.ofNullable(
-                    (Expression) process(node.getPartitionByColumn().get(),
-                        context))
-                    : Optional.empty()
-            )
-        );
+    return new CreateStreamAsSelect(
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java]: These are much cleaner! :D

**Comment by hjafarpour** [comment]
@big-andy-coates the `QueryTranslationTests` are coming in the next PR! :)

File: ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (miguno) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: I agree with @big-andy-coates that this is confusing.

> Can we not just fail if the table/stream name/alias matches any first-level column identifier?

This is one option to sort this out -- it's IMHO both simple plus easy to understand and remember (for the user ) and to explain (for us/KSQL error messages) --.  And I'd rather have this option than having to rely on users reading the docs (99% of the time they won't).

Open to other suggestions, of course, but I'd very much go with Andy's suggestion rather than with guessing what field would or would not be selected.

[the remark this answers is not in the document]
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: @big-andy-coates , @miguno Enforcing the requirement in DDL statemens to not have column name the same as stream name won't solve the problem. Users still cam use alias in their queries the same as column name and we will have to resolve this in query. Also in case of JOIN you can have column name from one side of the join be the same as the stream/table name from the other side and again we will  have ambiguity. 
I think the current behavior, defaulting to assume the stream name is the prefix is preferable.

[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Ah yes, it's more complex than I thought - thanks @hjafarpour.

The tricky one is the join. For join's you're joining two, or more, definitions together. I think its fine to fail a single command because its got ambiguities in it, but it's harder to fail something because the combination of multiple commands is leading to ambiguities. 

I don't _like_ the idea of not allowing certain stream / table names. But, without a better a approach, I think its the lesser of two evils.

```
CREATE STREAM X (Y STRUCT<Y INTEGER>, X INTEGER) ...;  -- valid
CREATE STREAM Y (X STRUCT<X INTEGER>, Y INTEGER) ...;  -- valid

SELECT X.X, Y.Y FROM X JOIN Y ON ...  -- very ambiguous!
```

One option, similar to what you're proposing, is to say that a struct path must always be fully qualified, i.e. it must always start with the table/stream name or alias.   This would avoid the ambiguity. The above would therefore become:

```
SELECT X.X, Y.Y FROM X JOIN Y ON ...  -- fails, as no suitable path can be found.

SELECT Y.X, Y.X.X, X.Y, X.Y,Y FROM X JOIN Y ON ... - explicit and valid.
```

I think its ok to say the path into a struct must always be fully qualified, as it avoids ambiguity.

The other alternative is to only fail where it is ambiguous.  So,

```
CREATE STREAM X (X STRUCT<X INTEGER>, Y STRUCT<Y INTEGER>) ...;  -- valid

SELECT Y.Y from X; -- is fine as its not ambiguous.
SELECT X.X.X from X; -- full qualified, so also fine.
SELECT X.X from X; -- fails as its ambiguous.
```

This is probably the nicest for the user. Though, if we support updating the upstream schemas we're in trouble. Consider the future, where KSQL doesn't require CS/CT statements, as it can just _discover_ streams and tables and their schemas in the cluster. Now we're in a situation where the schema of a source can evolve over time. This opens up the possibility of statements that once were unambiguous becoming ambiguous.

So... I think the safest option is to _require_ fully-qualified struct paths.

What say y'all?

[the remark this answers is not in the document]
Comment (dguy) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: I'm a +1 on using fully-qualified paths

[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: I think that might be the same as I was saying... maybe. Yes, 'X.X' would select the whole struct. But also, importantly, structs _always_ require a fully qualified path, even where it wouldn't be ambiguous to drop the table/stream name/alias e.g.

```
CREATE STREAM Bob (X STRUCT<X INTEGER>, Y STRUCT<Y INTEGER>) ...;
SELECT X.X FROM Bob; -- fails: Struct paths must be fully qualified.
```

The _benefit_ of this approach is that users will quickly come to _know_ that struct paths are always fully qualified. Therefore, they are much less likely to miss read something like:

```
CREATE STREAM X (X STRUCT<X INTEGER>) ...;
SELECT X.X FROM X;   -- selects the whole struct.
```

If we only require paths to be fully qualified when they are ambiguous users don't learn these nuances.

[the remark this answers is not in the document]
Comment (rodesai) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Interpreting the source name as referring to the source is more intuitive to me. The concept of resolving ambiguities with some precedence protocol should be familiar to people using computers.

That said, I'm +1 for requiring fully qualified paths for struct dereferences for the following reasons:

Like @big-andy-coates  said - requiring fully qualified paths for structs will trip users up more. This is both good and bad. A little bit bad because its not the most pleasant experience as you're getting started, but mostly good because with a decent error message its easy to learn and avoids confusion in the corner case where a column has the same name as the source.

Always interpreting the source name as referring to the source may break queries that work today, which we need to start taking seriously.

> if we support updating the upstream schemas we're in trouble. Consider the future, where KSQL doesn't require CS/CT statements, as it can just discover streams and tables and their schemas in the cluster. Now we're in a situation where the schema of a source can evolve over time. This opens up the possibility of statements that once were unambiguous becoming ambiguous.

This is an interesting point. I think for most other changes the schema-registry's compatibility checks can ensure that schemas don't change in an incompatible way. But they don't help for the join case.

[the remark this answers is not in the document]
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: I think here we are optimizing for a 1% case by imposing unnecessary requirement on 99% of the cases. We are also deviating from common practice of optional stream/table name/alias prefixing for columns. The only case we need to address ambiguity is when we have a situation like the following:

`CREATE STREAM A (A STRUCT<A INTEGER>, B STRUCT<B STRING>) WITH ...`

In the above case we have a stream, column and field of the column all having the same name. In every other case we won't have any ambiguity. The following behavior is intuitive and also does not require mandatory prefix for struct columns:

`SELECT A FROM A;         -- will return struct of column A`
`SELECT A.A FROM A;      -- will return struct of column A`
`SELECT A.A.A FROM A;   -- will return integer field with name A of the struct column A`
`SELECT B FROM A;           -- no ambiguity and no need for prefixing with table name`
`SELECT B.B FROM A;     -- again, no ambiguity and no need to enforce having table name prefix`

Enforcing users to always have stream/table name/alias as prefix for any struct type regardless of having any ambiguity would not create a positive user experience and I can imagine how many times users will be frustrated when getting error for not having the prefix.
I think we should also ask @rmoff and @blueedgenick 's opinion since they are more fluent in SQL and know our target users better.

[the remark this answers is not in the document]
Comment (apurvam) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: One concern with requiring full de-references for struct types is that you could run into potentially major inconveniences when the stream/table names are very long. For example, If you have a table named named `organization_activity_by_day` (a real example), then this can get pretty long pretty fast, especially if the struct field is verbose (and more so if individual column names are long).

That said, resolving the ambiguity problem using precedence rules is adds more cognitive overhead to the user. Similar looking statements would start meaning something different based on identifiers involved:

```
CREATE STREAM FOO (BAR STRUCT<X INTEGER,  ..>) WITH ..;
CREATE STREAM BAR (X bigint, Y bigint) WITH ...;

SELECT BAR.X, X FROM FOO JOIN BAR ON ...;
```

From what I have understood, this would select the same `X` (from `BAR`) twice if we use the suggested precedence rules, and is very non obvious. I think requiring struct types to be fully qualified will enable users to educate themselves about what to expect, albeit at the cost of some inconvenience.

On the balance, I think some inconvenience for a less loaded mental model is preferable, and am hence leaning towards requiring struct fields to be fully qualified.

[the remark this answers is not in the document]
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: @apurvam the example you have exactly shows that regardless of requiring or not requiring fully qualified path user needs to make the assumption that the stream/table name/alias has the precedence. If you enforce prefix on struct fields your will have the exact same query. With this assumption you can easily remove the ambiguity. 
So in both cases, the above query will return the same results.

Now let's assume user wants to select field `X` in the struct column `BAR` from stream `FOO` and also select the integer field `X` from stream `B`. For both mandatory and non mandatory prefix the queries will be the same:

``` SELECT FOO.BAR.X, X FROM FOO JOIN BAR ON ...;```

The main argument is that the query results for mandating and not mandating stream/table name/alias for struct fields is the same and mandating the prefix only adds extra burden  on user to alway add prefix even when it's not needed.

[the remark this answers is not in the document]
Comment (hjafarpour) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: Thanks for your feedback @rmoff . The problem is that even when we have ambiguity we won't be able show any error message. Consider the following example:

```CREATE STREAM A (A STRUCT<A INTEGER>, B STRUCT<B STRING>) WITH ...```
```SELECT A.A FROM A; --regardless of enforcing full path or not this won't fail and the result will be the same```

In the above example we will run the query and select the whole struct column and regardless of enforcing the full path or not it will always select the struct column instead of struct field. 
As for other SQL systems most of the traditional ones as you mentioned may not support nested types but most of the recent ones that are being used in big data community and streaming such as Hive, Impala, SparkSQL, and Calcite based systems (Flink, Phoenix, Kylin,...) do use dot notation and don't mandate the table name/alias prefixing for struct columns.

[the remark this answers is not in the document]
Comment (blueedgenick) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java]: adding another couple cents to the growing pile of pocket change: i happen to agree in large part with both @rmoff and @hjafarpour on this one :-) ..while it's true that most of the traditional RDBMS don't make this kind of addressing of nested structures very easy, with most of them resorting to function-based access, it's a matter of persepctive: they are all so old they never thought of having structured data as a type until much later, when it was hard to retrofit. So internally they mostly think of something like json as a varchar or clob that you can force "special rules" onto by use of a family of json-parsing functions. Similarly, in the past we thought that normalization and the assignment of attributes to tables was enough to handle any kind of "data structure" so `STRUCT` as a type made no sense. In the era of big-data systems it became much more common to allow access to json-like structures and `STRUCT`s natively, with a couple of prominent systems like Hive doing the same thing for avro data too. 
So, the questions of user expectations and familiarity might largely revolve around what you're used to, and i'd strongly favor going for the "big-data mindset" on this one, as it's closer to how we think in terms of native support for complex datatypes. 
I've had success in the past with avro data in Hive, where avro data translates neatly into a defined structure of nested maps, arrays, structs, and primitves, all of which have well-defined syntax for accessing them (with `[]`, `#` and `:` characters for array, map, and struct entries respectively). To query JSON in Hive, IIRC (worth double-checking this one, might have been improved upon!) you have to pre-declare a strucutre of maps/arrays/structs in the table schema, and then access as i just described for avro. So the query-writing becomes easy, and the question of dot-notation ambiguity never arises because of the use of different characters, but it often feels super clumsy to have to schematize your JSON this way.
Which brings us to Drill... not my favorite but it's focus on usability makes it possible to write dot-path access to nested fields within a query without first declaring any schema, which might be the closest to what we are looking for here ?
In summary, if i had to vote for one option, i'd probably go with something close to how Hive et al do this, for the largest community of folks who are already familiar with such an option, and therefore introduce the notion of different separator characters for different purposes. This works well for avro but note the trade-offs when querying JSON with this approach - it can still be done without declaring the structure ahead of time but the user needs to think of their json explicitly as a series of maps and arrays and then query it as such. I doubt that seems too foreign to any programmer types in the intended audience, the only question might be how intuitive it is to a more sql-background user.
